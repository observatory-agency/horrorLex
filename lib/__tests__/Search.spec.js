const Search = require('../Search');

describe('Search', () => {
  let search;
  const mockDocument = {
    title: 'book1',
  };
  const mockModel = {
    aggregate: jest.fn(() => ({
      toArray: jest.fn(() => [{
        documents: [mockDocument], total: [{ count: 100 }],
      }]),
    })),
    findOne: jest.fn(() => Promise.resolve(mockDocument)),
  };

  beforeEach(() => {
    search = new Search(mockModel);
  });

  describe('constructor', () => {
    it('should set this.model', () => {
      expect(search.model).toBe(mockModel);
    });
  });

  describe('one', () => {
    it('should return a document', async () => {
      const result = await search.one('some book slug');
      expect(result).toBe(mockDocument);
    });
  });

  describe('many', () => {
    describe('search', () => {
      describe('first page', () => {
        it('should return an object representing the results', async () => {
          const mockParams = {
            c: '10',
            p: '1',
            q: 'vampire',
            s: 'title',
            t: 'quick',
          };
          const results = await search.many(mockParams);
          expect(results).toMatchSnapshot();
        });
      });
      describe('last page', () => {
        it('should return an object representing the results', async () => {
          const mockParams = {
            c: '10',
            p: '100',
            q: 'vampire',
            s: 'title',
            t: 'quick',
          };
          const results = await search.many(mockParams);
          expect(results).toMatchSnapshot();
        });
      });
      describe('minimum amount of parameters', () => {
        it('should return an object representing the results', async () => {
          const mockParams = {
            q: 'vampire',
            t: 'quick',
          };
          const results = await search.many(mockParams);
          expect(results).toMatchSnapshot();
        });
      });
    });

    describe('tag', () => {
      describe('first page', () => {
        it('should return an object representing the results', async () => {
          const mockParams = {
            c: '10',
            p: '1',
            q: 'vampire',
            s: 'title',
            t: 'tag',
          };
          const results = await search.many(mockParams);
          expect(results).toMatchSnapshot();
        });
      });
      describe('last page', () => {
        it('should return an object representing the results', async () => {
          const mockParams = {
            c: '10',
            p: '100',
            q: 'vampire',
            s: 'title',
            t: 'tag',
          };
          const results = await search.many(mockParams);
          expect(results).toMatchSnapshot();
        });
      });
      describe('minimum amount of parameters', () => {
        it('should return an object representing the results', async () => {
          const mockParams = {
            q: 'vampire',
            t: 'tag',
          };
          const results = await search.many(mockParams);
          expect(results).toMatchSnapshot();
        });
      });
    });

    describe('no results', () => {
      beforeEach(() => {
        const mockModelNoResults = {
          aggregate: jest.fn(() => ({
            toArray: jest.fn(() => [{
              documents: [], total: [],
            }]),
          })),
        };
        search = new Search(mockModelNoResults);
      });
      it('should return "null"', async () => {
        const mockParams = {
          c: '10',
          p: '1',
          s: 'author',
          t: 'vampire',
        };
        const results = await search.many(mockParams);
        expect(results).toMatchSnapshot();
      });
    });
  });
});
