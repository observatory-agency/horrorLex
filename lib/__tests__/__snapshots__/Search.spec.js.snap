// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search many no results should return "null" 1`] = `null`;

exports[`Search many search first page should return an object representing the results 1`] = `
Object {
  "documents": Array [
    Object {
      "title": "book1",
    },
  ],
  "page": Object {
    "count": 10,
    "curr": 1,
    "next": 2,
    "number": 1,
    "pages": Array [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
    ],
    "params": "c=10&q=vampire&s=title&t=quick",
    "prev": 0,
    "query": "vampire",
    "range": 1,
    "skip": 0,
    "sort": "title",
    "total": 100,
    "type": "quick",
  },
  "total": Array [
    Object {
      "count": 100,
    },
  ],
}
`;

exports[`Search many search last page should return an object representing the results 1`] = `
Object {
  "documents": Array [
    Object {
      "title": "book1",
    },
  ],
  "page": Object {
    "count": 10,
    "curr": 100,
    "next": 0,
    "number": 100,
    "pages": Array [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
    ],
    "params": "c=10&q=vampire&s=title&t=quick",
    "prev": 99,
    "query": "vampire",
    "range": 991,
    "skip": 990,
    "sort": "title",
    "total": 100,
    "type": "quick",
  },
  "total": Array [
    Object {
      "count": 100,
    },
  ],
}
`;

exports[`Search many search minimum amount of parameters should return an object representing the results 1`] = `
Object {
  "documents": Array [
    Object {
      "title": "book1",
    },
  ],
  "page": Object {
    "count": 10,
    "curr": 1,
    "next": 2,
    "number": 1,
    "pages": Array [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
    ],
    "params": "q=vampire&t=quick&c=10&s=title",
    "prev": 0,
    "query": "vampire",
    "range": 1,
    "skip": 0,
    "sort": "title",
    "total": 100,
    "type": "quick",
  },
  "total": Array [
    Object {
      "count": 100,
    },
  ],
}
`;

exports[`Search many tag first page should return an object representing the results 1`] = `
Object {
  "documents": Array [
    Object {
      "title": "book1",
    },
  ],
  "page": Object {
    "count": 10,
    "curr": 1,
    "next": 2,
    "number": 1,
    "pages": Array [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
    ],
    "params": "c=10&q=vampire&s=title&t=tag",
    "prev": 0,
    "query": "vampire",
    "range": 1,
    "skip": 0,
    "sort": "title",
    "total": 100,
    "type": "tag",
  },
  "total": Array [
    Object {
      "count": 100,
    },
  ],
}
`;

exports[`Search many tag last page should return an object representing the results 1`] = `
Object {
  "documents": Array [
    Object {
      "title": "book1",
    },
  ],
  "page": Object {
    "count": 10,
    "curr": 100,
    "next": 0,
    "number": 100,
    "pages": Array [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
    ],
    "params": "c=10&q=vampire&s=title&t=tag",
    "prev": 99,
    "query": "vampire",
    "range": 991,
    "skip": 990,
    "sort": "title",
    "total": 100,
    "type": "tag",
  },
  "total": Array [
    Object {
      "count": 100,
    },
  ],
}
`;

exports[`Search many tag minimum amount of parameters should return an object representing the results 1`] = `
Object {
  "documents": Array [
    Object {
      "title": "book1",
    },
  ],
  "page": Object {
    "count": 10,
    "curr": 1,
    "next": 2,
    "number": 1,
    "pages": Array [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
    ],
    "params": "q=vampire&t=tag&c=10&s=title",
    "prev": 0,
    "query": "vampire",
    "range": 1,
    "skip": 0,
    "sort": "title",
    "total": 100,
    "type": "tag",
  },
  "total": Array [
    Object {
      "count": 100,
    },
  ],
}
`;
